crates/ide_assists/Cargo.toml

```diff
@@ -19,6 +19,7 @@ syntax = { path = "../syntax", version = "0.0.0" }
 text_edit = { path = "../text_edit", version = "0.0.0" }
 profile = { path = "../profile", version = "0.0.0" }
 ide_db = { path = "../ide_db", version = "0.0.0" }
+ide_ssr = { path = "../ide_ssr", version = "0.0.0" }
```

Comment:
I am not sure that doing this via a literal assist in this crate is the best thing. ssr doesn't look like other assists,  and it'd be cool to compile assists and ssr in parallel. 

So perhaps this could be done in the `ide/` layer? Either ther's a separat call to check/apply assist, or else an IDE can smuggle an unusual assist into the `assists` array:

```
    /// Computes assists (aka code actions aka intentions) for the given
    /// position. If `resolve == false`, computes enough info to show the
    /// lightbulb list in the editor, but doesn't compute actual edits, to
    /// improve performance.
    pub fn assists(
        &self,
        config: &AssistConfig,
        resolve: bool,
        frange: FileRange,
    ) -> Cancelable<Vec<Assist>> {
        self.with_db(|db| { let mut res = Assist::get(db, config, resolve, frange);  res.push(ssr)})
    }
```

Reply:
Thanks for the feedback!

> Either ther's a separat call to check/apply assist

I'm not sure I understand what you mean here.

> or else an IDE can smuggle an unusual assist into the assists array

I could see a path forward here. But I think it would involve opening up the public interface of the `ide_assists` crate quite a bit.  

For example, right now `AssistContext` is `pub(crate)`. Plus all the assist test code would either need to be duplicated into the `ide` crate or made public. 

Is this okay? It seems to me like a big architectural cost with the benefit being only that we wouldn't have to compile `ide_ssr` before `ide_assists`.

Reply:
> For example, right now AssistContext is pub(crate). Plus all the assist test infrastructure would either need to be duplicated into the ide crate or made public.

I don’t thing you actually need assist infra? It is there to do code analysis and editing, but for us all is already handled by ssr. 

What we need is only ability to create an Assist, and you should be able to do that manually, bypassing AssistCtx, as assist itself is just data.

Reply:
Ah, that makes sense. I was thinking I needed to somehow slip an assist handler into the list of handlers. But yeah I think I see now that I can just hand-craft an instance of the `Assist` struct from within the `ide` crate. I'll give that a shot!

Reply:
Alright so I've re-implemented this in the `ide` crate, just manually creating the `Assist` struct. There is some issue with it, so it doesn't actually work. But before I dig into that I still have some architectural concerns with duplicating assist logic. I don't know much about this, so let me know if these concerns aren't valid.

1. Anything that `AssistConfig` is used for would need to be reimplemented.
2. This new approach doesn't provide any possibility of splitting the work of determining if an assist is applicable vs calculating the edit for the assist.

My current feeling is that this should really be a UI question. Do we want to provide this as an assist or as a new command? I'd personally prefer it as an assist because it happens to suit my workflow. But I think there may be some other points in favor of it being an assist, such as there are no arguments to this function, other than your current cursor position, which matches the model for an assist. 

I'm not familiar enough with LSP to know, but I believe implementing this as a command rather than an assist involves additional code in rust-analyzer (something in `rust-analyzer/src/handlers.rs`), plus supporting code in any lsp clients (similar to [here](https://github.com/fannheyward/coc-rust-analyzer/blob/master/src/commands.ts#L139)). 

Could you clarify your concerns about having this as an assist? Is it mostly a UI concern, or mostly a code architecture / compile time concern, or something else?

Reply:
I think you're right about the command vs. assist thing. Commands are custom, while assists work in every client. Since we can't realistically maintain more than a couple of client plugins, an assist would be better. In addition, in the long run we want to upstream our LSP protocol extensions, reducing the amount of client-specific code. And it's probably not going to happen soon, but you could also imagine syntax highlighting and completion being available in the SSR comment.

The disadvantage of an assist is discoverability. You can see a list of `rust-analyzer` commands, but there's no way you're going to run into the SSR assist without reading the manual.

Reply:
I do think UI-wise, it should be assist. I don't think that the implementation belongs to the assits module, as that introduce dep edge between assist crate (which is big) and ssr crate (which is big). So we should achieve the UI we want without changing the physical architecture. I think, the following should do the trick:

```diff
diff --git a/crates/ide/src/lib.rs b/crates/ide/src/lib.rs
index b600178ee..769069191 100644
--- a/crates/ide/src/lib.rs
+++ b/crates/ide/src/lib.rs
@@ -504,7 +504,11 @@ impl Analysis {
         resolve: bool,
         frange: FileRange,
     ) -> Cancelable<Vec<Assist>> {
-        self.with_db(|db| Assist::get(db, config, resolve, frange))
+        self.with_db(|db| {
+            let mut acc = Assist::get(db, config, resolve, frange);
+            add_ssr(db, &mut acc, resolve, frange);
+            acc
+        })
     }
 
     /// Computes the set of diagnostics for the given file.
@@ -579,6 +583,42 @@ impl Analysis {
     }
 }
 
+fn add_ssr(
+    db: &RootDatabase,
+    base: &mut Vec<Assist>,
+    resolve: bool,
+    frange: FileRange,
+) -> Option<()> {
+    use syntax::ast::{self, AstNode, AstToken};
+
+    let comment = {
+        let file = db.parse(frange.file_id);
+        file.tree().syntax().token_at_offset(frange.range.start()).find_map(ast::Comment::cast)
+    }?;
+
+    let comment_text_without_prefix = comment.text().strip_prefix(comment.prefix()).unwrap();
+    // Attempt to parse the comment line as an SSR rule. If the parsing fails
+    // this assist is not applicable.
+    let ssr_rule = comment_text_without_prefix.parse().ok()?;
+    let lookup_context = FilePosition { file_id: frange.file_id, offset: frange.range.start() };
+    let mut match_finder = ide_ssr::MatchFinder::in_context(db, lookup_context, vec![]);
+    match_finder.add_rule(ssr_rule).ok()?;
+
+    let mut assist = Assist {
+        id: AssistId("ssr", AssistKind::RefactorRewrite),
+        label: Label::new("Apply SSR"),
+        group: None,
+        target: comment.syntax().text_range(),
+        source_change: None,
+    };
+    if resolve {
+        assist.source_change = Some(SourceChange::from_edits(match_finder.edits(), Vec::new()));
+    }
+
+    base.push(assist);
+    Some(())
+}
+
 #[test]
 fn analysis_is_send() {
     fn is_send<T: Send>() {}
```

(with maybe moving the actual logic here to an `ssr_from_comment` function insde ssr)

Reply:
Thanks for the detailed response! I've taken this suggestion and the code is quite different now so I'm going to go ahead and resolve this one and open a few new discussions for follow up questions I have.